diff --git a/tools/rust_analyzer/BUILD.bazel b/tools/rust_analyzer/BUILD.bazel
index 5adc63b..4d8c7a1 100644
--- a/tools/rust_analyzer/BUILD.bazel
+++ b/tools/rust_analyzer/BUILD.bazel
@@ -1,3 +1,6 @@
+load("//rust:defs.bzl", "rust_binary")
+load(":opt_transition.bzl", "opt_executable")
+
 package(default_visibility = ["//visibility:public"])
 
 exports_files([
@@ -48,3 +51,78 @@ alias(
     actual = "@rules_rs//tools/rust_analyzer:bzl_lib",
     tags = ["manual"],
 )
+
+rust_binary(
+    name = "flycheck",
+    srcs = ["bin/flycheck.rs"],
+    edition = "2021",
+    deps = [
+        ":gen_rust_project_lib",
+        "@rrra//:anyhow",
+        "@rrra//:camino",
+        "@rrra//:clap",
+        "@rrra//:env_logger",
+        "@rrra//:log",
+        "@rrra//:serde_json",
+    ],
+)
+
+opt_executable(
+    name = "discover_bazel_rust_project_opt",
+    src = ":discover_bazel_rust_project",
+    visibility = ["//visibility:private"],
+)
+
+opt_executable(
+    name = "flycheck_opt",
+    src = ":flycheck",
+    visibility = ["//visibility:private"],
+)
+
+rust_binary(
+    name = "launcher",
+    srcs = ["bin/launcher.rs"],
+    edition = "2021",
+    deps = [
+        "@rrra//:serde_json",
+    ],
+)
+
+opt_executable(
+    name = "launcher_opt",
+    src = ":launcher",
+    visibility = ["//visibility:private"],
+)
+
+rust_binary(
+    name = "setup",
+    srcs = ["bin/setup.rs"],
+    data = [
+        ":discover_bazel_rust_project_opt",
+        ":flycheck_opt",
+        ":launcher_opt",
+        "//rust/toolchain:current_rust_analyzer_toolchain",
+        "//rust/toolchain:current_rustfmt_toolchain",
+    ],
+    edition = "2021",
+    rustc_env = {
+        "LAUNCHER_RLOCATIONPATH": "$(rlocationpath :launcher_opt)",
+        "RUSTFMT_RLOCATIONPATH": "$(RUSTFMT_RLOCATIONPATH)",
+        "RUST_ANALYZER_PROC_MACRO_SRV_RLOCATIONPATH": "$(RUST_ANALYZER_PROC_MACRO_SRV_RLOCATIONPATH)",
+        "RUST_ANALYZER_RLOCATIONPATH": "$(RUST_ANALYZER_RLOCATIONPATH)",
+    },
+    toolchains = [
+        "//rust/toolchain:current_rust_analyzer_toolchain",
+        "//rust/toolchain:current_rustfmt_toolchain",
+    ],
+    deps = [
+        ":gen_rust_project_lib",
+        "//rust/runfiles",
+        "@rrra//:anyhow",
+        "@rrra//:camino",
+        "@rrra//:clap",
+        "@rrra//:env_logger",
+        "@rrra//:log",
+        "@rrra//:serde_json",
+    ],
+)
